File: gopost_app/lib/video_editor/presentation/providers/media_pool_notifier.dart
import 'dart:io';

import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:gopost_app/video_editor/domain/models/media_asset.dart';

/// Not autoDispose — the Media Pool state must survive panel switches.
/// The EditorPanelArea mounts/unmounts panel widgets on tab change, so
/// autoDispose would wipe all imported assets whenever the user navigates
/// to a different sidebar tab and back.
final mediaPoolNotifierProvider =
    StateNotifierProvider<MediaPoolNotifier, MediaPoolState>((ref) {
  return MediaPoolNotifier();
});

enum MediaPoolViewMode { grid, list }

enum MediaPoolSortBy { name, date, type, duration, size }

class MediaPoolState {
  final List<MediaAsset> assets;
  final List<MediaBin> bins;
  final String? activeBinId;
  final String searchQuery;
  final MediaAssetType? filterType;
  final MediaPoolViewMode viewMode;
  final MediaPoolSortBy sortBy;
  final bool sortAscending;
  final bool isImporting;
  final int importTotal;
  final int importDone;

  const MediaPoolState({
    this.assets = const [],
    this.bins = const [],
    this.activeBinId,
    this.searchQuery = '',
    this.filterType,
    this.viewMode = MediaPoolViewMode.grid,
    this.sortBy = MediaPoolSortBy.date,
    this.sortAscending = false,
    this.isImporting = false,
    this.importTotal = 0,
    this.importDone = 0,
  });

  /// Assets filtered by current bin, search, and type filter.
  List<MediaAsset> get filteredAssets {
    var result = assets.where((a) {
      if (activeBinId != null && a.binId != activeBinId) return false;
      if (activeBinId == null && a.binId != null) {
        // Show all assets when no bin is selected (root view)
      }
      if (filterType != null && a.type != filterType) return false;
      if (searchQuery.isNotEmpty) {
        final q = searchQuery.toLowerCase();
        return a.fileName.toLowerCase().contains(q) ||
            a.codec.toLowerCase().contains(q);
      }
      return true;
    }).toList();

    result.sort((a, b) {
      int cmp;
      switch (sortBy) {
        case MediaPoolSortBy.name:
          cmp = a.fileName.toLowerCase().compareTo(b.fileName.toLowerCase());
        case MediaPoolSortBy.date:
          cmp = a.importedAt.compareTo(b.importedAt);
        case MediaPoolSortBy.type:
          cmp = a.type.index.compareTo(b.type.index);
        case MediaPoolSortBy.duration:
          cmp = a.durationSeconds.compareTo(b.durationSeconds);
        case MediaPoolSortBy.size:
          cmp = a.fileSizeBytes.compareTo(b.fileSizeBytes);
      }
      return sortAscending ? cmp : -cmp;
    });

    return result;
  }

  int get assetCount => assets.length;
  int get offlineCount => assets.where((a) => a.status == MediaAssetStatus.offline).length;

  MediaPoolState copyWith({
    List<MediaAsset>? assets,
    List<MediaBin>? bins,
    String? activeBinId,
    bool clearActiveBin = false,
    String? searchQuery,
    MediaAssetType? filterType,
    bool clearFilter = false,
    MediaPoolViewMode? viewMode,
    MediaPoolSortBy? sortBy,
    bool? sortAscending,
    bool? isImporting,
    int? importTotal,
    int? importDone,
  }) {
    return MediaPoolState(
      assets: assets ?? this.assets,
      bins: bins ?? this.bins,
      activeBinId: clearActiveBin ? null : (activeBinId ?? this.activeBinId),
      searchQuery: searchQuery ?? this.searchQuery,
      filterType: clearFilter ? null : (filterType ?? this.filterType),
      viewMode: viewMode ?? this.viewMode,
      sortBy: sortBy ?? this.sortBy,
      sortAscending: sortAscending ?? this.sortAscending,
      isImporting: isImporting ?? this.isImporting,
      importTotal: importTotal ?? this.importTotal,
      importDone: importDone ?? this.importDone,
    );
  }
}

class MediaPoolNotifier extends StateNotifier<MediaPoolState> {
  MediaPoolNotifier() : super(const MediaPoolState());

  static const _videoExts = {'.mp4', '.mov', '.avi', '.mkv', '.webm', '.m4v', '.flv', '.wmv', '.3gp'};
  static const _imageExts = {'.jpg', '.jpeg', '.png', '.webp', '.gif', '.bmp', '.heic', '.heif', '.tiff'};
  static const _audioExts = {'.mp3', '.aac', '.wav', '.m4a', '.flac', '.ogg', '.wma', '.opus', '.aiff'};

  static MediaAssetType classifyFile(String path) {
    final ext = '.${path.toLowerCase().split('.').last}';
    if (_videoExts.contains(ext)) return MediaAssetType.video;
    if (_imageExts.contains(ext)) return MediaAssetType.image;
    if (_audioExts.contains(ext)) return MediaAssetType.audio;
    return MediaAssetType.video;
  }

  /// Import a single file into the Media Pool.
  /// Returns the created [MediaAsset] or null on failure.
  Future<MediaAsset?> importFile(
    String filePath, {
    String? binId,
    int width = 0,
    int height = 0,
    double durationSeconds = 0,
    double frameRate = 0,
    String codec = '',
  }) async {
    // Deduplicate by file path
    if (state.assets.any((a) => a.filePath == filePath)) {
      return state.assets.firstWhere((a) => a.filePath == filePath);
    }

    final file = File(filePath);
    final exists = await file.exists();
    int fileSize = 0;
    if (exists) {
      try {
        fileSize = await file.length();
      } catch (_) {}
    }

    final fileName = filePath.split(RegExp(r'[/\\]')).last;
    final type = classifyFile(filePath);
    final id = '${DateTime.now().millisecondsSinceEpoch}_${fileName.hashCode.abs()}';

    final asset = MediaAsset(
      id: id,
      filePath: filePath,
      fileName: fileName,
      type: type,
      status: exists ? MediaAssetStatus.online : MediaAssetStatus.offline,
      binId: binId,
      width: width,
      height: height,
      durationSeconds: durationSeconds,
      frameRate: frameRate,
      codec: codec,
      fileSizeBytes: fileSize,
      importedAt: DateTime.now(),
    );

    state = state.copyWith(assets: [...state.assets, asset]);
    return asset;
  }

  /// Update an asset's metadata (e.g. after probing).
  void updateAssetMetadata(
    String assetId, {
    int? width,
    int? height,
    double? durationSeconds,
    double? frameRate,
    String? codec,
    int? fileSizeBytes,
  }) {
    state = state.copyWith(
      assets: state.assets.map((a) {
        if (a.id != assetId) return a;
        return a.copyWith(
          width: width,
          height: height,
          durationSeconds: durationSeconds,
          frameRate: frameRate,
          codec: codec,
          fileSizeBytes: fileSizeBytes,
        );
      }).toList(),
    );
  }

  /// Remove an asset from the pool.
  void removeAsset(String assetId) {
    state = state.copyWith(
      assets: state.assets.where((a) => a.id != assetId).toList(),
    );
  }

  /// Move an asset to a bin (null = root).
  void moveAssetToBin(String assetId, String? binId) {
    state = state.copyWith(
      assets: state.assets.map((a) {
        if (a.id != assetId) return a;
        return a.copyWith(binId: binId, clearBin: binId == null);
      }).toList(),
    );
  }

  // ---------------------------------------------------------------------------
  // Bins
  // ---------------------------------------------------------------------------

  void createBin(String name, {String? parentId}) {
    final id = 'bin_${DateTime.now().millisecondsSinceEpoch}';
    final bin = MediaBin(id: id, name: name, parentId: parentId);
    state = state.copyWith(bins: [...state.bins, bin]);
  }

  void renameBin(String binId, String newName) {
    state = state.copyWith(
      bins: state.bins.map((b) {
        if (b.id != binId) return b;
        return b.copyWith(name: newName);
      }).toList(),
    );
  }

  void deleteBin(String binId) {
    // Move assets from deleted bin back to root
    state = state.copyWith(
      bins: state.bins.where((b) => b.id != binId).toList(),
      assets: state.assets.map((a) {
        if (a.binId != binId) return a;
        return a.copyWith(clearBin: true);
      }).toList(),
      clearActiveBin: state.activeBinId == binId,
    );
  }

  void setActiveBin(String? binId) {
    state = state.copyWith(
      activeBinId: binId,
      clearActiveBin: binId == null,
    );
  }

  // ---------------------------------------------------------------------------
  // Filters & view
  // ---------------------------------------------------------------------------

  void setSearchQuery(String query) {
    state = state.copyWith(searchQuery: query);
  }

  void setFilterType(MediaAssetType? type) {
    state = state.copyWith(
      filterType: type,
      clearFilter: type == null,
    );
  }

  void setViewMode(MediaPoolViewMode mode) {
    state = state.copyWith(viewMode: mode);
  }

  void setSortBy(MediaPoolSortBy sort) {
    if (state.sortBy == sort) {
      state = state.copyWith(sortAscending: !state.sortAscending);
    } else {
      state = state.copyWith(sortBy: sort, sortAscending: true);
    }
  }

  void setImportProgress({required bool importing, int total = 0, int done = 0}) {
    state = state.copyWith(
      isImporting: importing,
      importTotal: total,
      importDone: done,
    );
  }

  // ---------------------------------------------------------------------------
  // Offline / relink
  // ---------------------------------------------------------------------------

  /// Check all assets for offline status.
  Future<void> checkAllAssetsOnline() async {
    final updated = <MediaAsset>[];
    for (final asset in state.assets) {
      final online = await asset.checkOnline();
      updated.add(asset.copyWith(
        status: online ? MediaAssetStatus.online : MediaAssetStatus.offline,
      ));
    }
    state = state.copyWith(assets: updated);
  }

  /// Relink an offline asset to a new file path.
  Future<bool> relinkAsset(String assetId, String newPath) async {
    final file = File(newPath);
    if (!await file.exists()) return false;

    int fileSize = 0;
    try {
      fileSize = await file.length();
    } catch (_) {}

    state = state.copyWith(
      assets: state.assets.map((a) {
        if (a.id != assetId) return a;
        return a.copyWith(
          filePath: newPath,
          status: MediaAssetStatus.online,
          fileSizeBytes: fileSize,
        );
      }).toList(),
    );
    return true;
  }

  // ---------------------------------------------------------------------------
  // Persistence
  // ---------------------------------------------------------------------------

  /// Load pool data (called when project opens).
  void loadFromData(MediaPoolData data) {
    state = state.copyWith(
      assets: data.assets,
      bins: data.bins,
    );
  }

  /// Export pool data for project save.
  MediaPoolData toData() {
    return MediaPoolData(
      assets: state.assets,
      bins: state.bins,
    );
  }
}

File: gopost_app_qt/src/video_editor/presentation/notifiers/media_pool_notifier.cpp
#include "video_editor/presentation/notifiers/media_pool_notifier.h"

#include <QFileInfo>
#include <QUrl>
#include <QUuid>
#include <QVariantMap>
#include <algorithm>

namespace gopost::video_editor {

// ---------------------------------------------------------------------------
// Filtered assets
// ---------------------------------------------------------------------------

std::vector<const MediaAsset*> MediaPoolState::filteredAssets() const {
    std::vector<const MediaAsset*> result;
    for (const auto& a : assets) {
        // Bin filter
        if (activeBinId && a.binId != *activeBinId) continue;
        // Type filter
        if (filterType.has_value() && a.type != *filterType) continue;
        // Search
        if (!searchQuery.isEmpty() &&
            !a.fileName.contains(searchQuery, Qt::CaseInsensitive)) continue;
        result.push_back(&a);
    }

    // Sort
    auto cmp = [this](const MediaAsset* a, const MediaAsset* b) -> bool {
        int c = 0;
        switch (sortBy) {
        case MediaPoolSortBy::Name:     c = a->fileName.compare(b->fileName, Qt::CaseInsensitive); break;
        case MediaPoolSortBy::Date:     c = a->importedAt < b->importedAt ? -1 : (a->importedAt > b->importedAt ? 1 : 0); break;
        case MediaPoolSortBy::Type:     c = static_cast<int>(a->type) - static_cast<int>(b->type); break;
        case MediaPoolSortBy::Duration: c = a->durationSeconds < b->durationSeconds ? -1 : (a->durationSeconds > b->durationSeconds ? 1 : 0); break;
        case MediaPoolSortBy::Size:     c = a->fileSizeBytes < b->fileSizeBytes ? -1 : (a->fileSizeBytes > b->fileSizeBytes ? 1 : 0); break;
        }
        return sortAscending ? c < 0 : c > 0;
    };
    std::sort(result.begin(), result.end(), cmp);
    return result;
}

// ---------------------------------------------------------------------------

MediaPoolNotifier::MediaPoolNotifier(QObject* parent) : QObject(parent) {}
MediaPoolNotifier::~MediaPoolNotifier() = default;

QString MediaPoolNotifier::generateAssetId() const {
    return QUuid::createUuid().toString(QUuid::WithoutBraces);
}

// ---------------------------------------------------------------------------
// Assets
// ---------------------------------------------------------------------------

void MediaPoolNotifier::importFile(const QString& path) {
    qDebug() << "[MediaPool] importFile:" << path;
    QFileInfo fi(path);
    MediaAsset asset;
    asset.id            = generateAssetId();
    asset.filePath      = path;
    asset.fileName      = fi.fileName();
    asset.fileSizeBytes = fi.size();
    asset.importedAt    = QDateTime::currentDateTime();
    asset.status        = fi.exists() ? MediaAssetStatus::Online : MediaAssetStatus::Offline;

    // Detect type by extension
    const auto ext = fi.suffix().toLower();
    static const QStringList videoExts = {
        QStringLiteral("mp4"), QStringLiteral("mov"), QStringLiteral("avi"),
        QStringLiteral("mkv"), QStringLiteral("webm"), QStringLiteral("m4v"),
        QStringLiteral("flv"), QStringLiteral("wmv"), QStringLiteral("3gp")
    };
    static const QStringList imageExts = {
        QStringLiteral("jpg"), QStringLiteral("jpeg"), QStringLiteral("png"),
        QStringLiteral("webp"), QStringLiteral("gif"), QStringLiteral("bmp"),
        QStringLiteral("heic"), QStringLiteral("heif"), QStringLiteral("tiff")
    };
    static const QStringList audioExts = {
        QStringLiteral("mp3"), QStringLiteral("aac"), QStringLiteral("wav"),
        QStringLiteral("m4a"), QStringLiteral("flac"), QStringLiteral("ogg"),
        QStringLiteral("wma"), QStringLiteral("opus"), QStringLiteral("aiff")
    };

    if (videoExts.contains(ext))       asset.type = MediaAssetType::Video;
    else if (imageExts.contains(ext))  asset.type = MediaAssetType::Image;
    else if (audioExts.contains(ext))  asset.type = MediaAssetType::Audio;
    else                               asset.type = MediaAssetType::Video; // default

    state_.assets.push_back(std::move(asset));
    emit stateChanged();
}

void MediaPoolNotifier::removeAsset(const QString& assetId) {
    auto& a = state_.assets;
    a.erase(std::remove_if(a.begin(), a.end(),
                            [&](const MediaAsset& m) { return m.id == assetId; }),
            a.end());
    emit stateChanged();
}

void MediaPoolNotifier::moveAssetToBin(const QString& assetId, const QString& binId) {
    for (auto& a : state_.assets) {
        if (a.id == assetId) { a.binId = binId; break; }
    }
    emit stateChanged();
}

void MediaPoolNotifier::relinkAsset(const QString& assetId, const QString& newPath) {
    for (auto& a : state_.assets) {
        if (a.id == assetId) {
            a.filePath = newPath;
            a.status   = QFileInfo::exists(newPath) ? MediaAssetStatus::Online
                                                     : MediaAssetStatus::Offline;
            break;
        }
    }
    emit stateChanged();
}

void MediaPoolNotifier::checkAllAssetsOnline() {
    for (auto& a : state_.assets)
        a.status = a.checkOnline() ? MediaAssetStatus::Online : MediaAssetStatus::Offline;
    emit stateChanged();
}

// ---------------------------------------------------------------------------
// Bins
// ---------------------------------------------------------------------------

void MediaPoolNotifier::createBin(const QString& name) {
    MediaBin bin;
    bin.id   = generateAssetId();
    bin.name = name;
    state_.bins.push_back(std::move(bin));
    emit stateChanged();
}

void MediaPoolNotifier::renameBin(const QString& binId, const QString& name) {
    for (auto& b : state_.bins)
        if (b.id == binId) { b.name = name; break; }
    emit stateChanged();
}

void MediaPoolNotifier::removeBin(const QString& binId) {
    auto& b = state_.bins;
    b.erase(std::remove_if(b.begin(), b.end(),
                            [&](const MediaBin& m) { return m.id == binId; }),
            b.end());
    if (state_.activeBinId == binId)
        state_.activeBinId.reset();
    emit stateChanged();
}

void MediaPoolNotifier::setActiveBin(const QString& binId) {
    state_.activeBinId = binId.isEmpty() ? std::nullopt : std::optional(binId);
    emit stateChanged();
}

// ---------------------------------------------------------------------------
// QML property helpers
// ---------------------------------------------------------------------------

QVariantList MediaPoolNotifier::filteredAssetsVariant() const {
    QVariantList result;
    auto filtered = state_.filteredAssets();
    for (const auto* asset : filtered) {
        QVariantMap m;
        m[QStringLiteral("id")] = asset->id;
        m[QStringLiteral("displayName")] = asset->fileName;
        m[QStringLiteral("filePath")] = asset->filePath;
        m[QStringLiteral("mediaType")] = [&] {
            switch (asset->type) {
            case MediaAssetType::Video: return QStringLiteral("video");
            case MediaAssetType::Image: return QStringLiteral("image");
            case MediaAssetType::Audio: return QStringLiteral("audio");
            default:                    return QStringLiteral("video");
            }
        }();
        m[QStringLiteral("duration")] = asset->durationSeconds;
        m[QStringLiteral("fileSize")] = static_cast<qint64>(asset->fileSizeBytes);
        m[QStringLiteral("isSelected")] = (selectedAssetId_ == asset->id);
        result.append(m);
    }
    return result;
}

QString MediaPoolNotifier::currentFilter() const {
    if (!state_.filterType.has_value()) return QStringLiteral("all");
    switch (*state_.filterType) {
    case MediaAssetType::Video: return QStringLiteral("video");
    case MediaAssetType::Image: return QStringLiteral("image");
    case MediaAssetType::Audio: return QStringLiteral("audio");
    default:                    return QStringLiteral("all");
    }
}

void MediaPoolNotifier::selectAsset(const QString& assetId) {
    selectedAssetId_ = assetId;
    emit stateChanged();
}

void MediaPoolNotifier::addToTimeline(const QString& assetId) {
    qDebug() << "[MediaPool] addToTimeline called with id:" << assetId
             << "total assets:" << state_.assets.size();
    for (const auto& a : state_.assets) {
        if (a.id == assetId) {
            QString mediaType;
            switch (a.type) {
            case MediaAssetType::Video: mediaType = QStringLiteral("video"); break;
            case MediaAssetType::Image: mediaType = QStringLiteral("image"); break;
            case MediaAssetType::Audio: mediaType = QStringLiteral("audio"); break;
            default: mediaType = QStringLiteral("video"); break;
            }
            qDebug() << "[MediaPool] emitting addToTimelineRequested:"
                     << a.filePath << mediaType << a.fileName << a.durationSeconds;
            emit addToTimelineRequested(assetId, a.filePath, mediaType,
                                        a.fileName, a.durationSeconds);
            return;
        }
    }
    qWarning() << "[MediaPool] addToTimeline: asset NOT FOUND for id:" << assetId;
}

void MediaPoolNotifier::showFilePicker() {
    emit filePickerRequested();
}

void MediaPoolNotifier::importFiles(const QVariantList& urls) {
    for (const auto& urlVar : urls) {
        // Qt6 QML may double-wrap QUrl inside QVariant(QVariant(QUrl)).
        // Unwrap until we get a usable type.
        QVariant v = urlVar;
        while (v.typeId() == QMetaType::QVariant)
            v = v.value<QVariant>();

        QString path;
        if (v.canConvert<QUrl>()) {
            path = v.toUrl().toLocalFile();
        } else {
            path = v.toString();
            if (path.startsWith(QStringLiteral("file:///")))
                path = QUrl(path).toLocalFile();
            else if (path.startsWith(QStringLiteral("file://")))
                path = QUrl(path).toLocalFile();
        }
        qDebug() << "[MediaPool] importFiles resolved:" << path;
        if (!path.isEmpty())
            importFile(path);
    }
}

void MediaPoolNotifier::setFilter(const QString& filter) {
    if (filter == QStringLiteral("all") || filter.isEmpty()) {
        clearFilterType();
    } else if (filter == QStringLiteral("video")) {
        setFilterType(static_cast<int>(MediaAssetType::Video));
    } else if (filter == QStringLiteral("image")) {
        setFilterType(static_cast<int>(MediaAssetType::Image));
    } else if (filter == QStringLiteral("audio")) {
        setFilterType(static_cast<int>(MediaAssetType::Audio));
    }
}

void MediaPoolNotifier::cycleSort() {
    int current = static_cast<int>(state_.sortBy);
    int next = (current + 1) % 5; // 5 sort modes
    state_.sortBy = static_cast<MediaPoolSortBy>(next);
    emit stateChanged();
}

// ---------------------------------------------------------------------------
// Search / filter / sort
// ---------------------------------------------------------------------------

void MediaPoolNotifier::setSearchQuery(const QString& query) {
    state_.searchQuery = query;
    emit stateChanged();
}

void MediaPoolNotifier::setFilterType(int type) {
    state_.filterType = static_cast<MediaAssetType>(type);
    emit stateChanged();
}

void MediaPoolNotifier::clearFilterType() {
    state_.filterType.reset();
    emit stateChanged();
}

void MediaPoolNotifier::setSortBy(int sortBy) {
    state_.sortBy = static_cast<MediaPoolSortBy>(sortBy);
    emit stateChanged();
}

void MediaPoolNotifier::toggleSortOrder() {
    state_.sortAscending = !state_.sortAscending;
    emit stateChanged();
}

void MediaPoolNotifier::setViewMode(int mode) {
    state_.viewMode = static_cast<MediaPoolViewMode>(mode);
    emit stateChanged();
}

// ---------------------------------------------------------------------------
// Persistence
// ---------------------------------------------------------------------------

void MediaPoolNotifier::loadFromData(const QVariantList& data) {
    Q_UNUSED(data);
    // TODO: Deserialise from QVariantList
}

QVariantList MediaPoolNotifier::toData() const {
    QVariantList result;
    // TODO: Serialise to QVariantList
    return result;
}

} // namespace gopost::video_editor
